math/src/polynom/multi_linear.rs

```diff
@@ -0,0 +1,244 @@
+/// evaluation basis for evaluation set ${0 , 1}^ν$.
+///
+/// Given a function $(f: {0 , 1}^ν ⇾ 𝔽)$, its multi-linear extension (i.e., the unique
+/// mult-linear polynomial extending `f` to $(f^{~}: 𝔽^ν ⇾ 𝔽)$ and agreeing with it on ${0 , 1}^ν$) is
+/// defined as the summation of the evaluations of f against the Lagrange basis.
+/// More specifically, given $(r_0, ..., r_{ν - 1}) ∈ 𝔽^ν$, then:
+///
+/// $$
+///     f^{~}(r_0, ..., r_{ν - 1}) = \sum_{(y_0, ..., y_{ν - 1}) ∈ {0 , 1}^ν}
+///                  f(y_0, ..., y_{ν - 1}) EQ^{~}((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1}))
+/// $$
+///
+/// We call the Lagrange kernel the evaluation of the EQ^{~} function at
+/// $((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1}))$ for all $(y_0, ..., y_{ν - 1}) ∈ {0 , 1}^ν$ for
+/// a fixed $(r_0, ..., r_{ν - 1}) ∈ 𝔽^ν$.
+///
+/// [`EqFunction`] represents EQ^{~} the multi-linear extension of
+///
+/// $((y_0, ..., y_{ν - 1}) ↦ EQ((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1})))$
+///
+/// and contains a method to generate the Lagrange kernel for defining evaluations of multi-linear
+/// extensions of arbitrary functions $(f: {0 , 1}^ν ⇾ 𝔽)$ at a given point $(r_0, ..., r_{ν - 1})$
+/// as well as a method to evaluate $EQ^{~}((r_0, ..., r_{ν - 1}), (t_0, ..., t_{ν - 1})))$ for
+/// $(t_0, ..., t_{ν - 1}) ∈ 𝔽^ν$.
+pub struct EqFunction<E> {
+    r: Vec<E>,
+}
+
+impl<E: FieldElement> EqFunction<E> {
+    /// Creates a new [EqFunction].
+    pub fn new(r: Vec<E>) -> Self {
+        let tmp = r.clone();
+        EqFunction { r: tmp }
+    }
+
+    /// Computes $EQ((r_0, ..., r_{ν - 1}), (t_0, ..., t_{ν - 1})))$.
+    pub fn evaluate(&self, t: &[E]) -> E {
+        assert_eq!(self.r.len(), t.len());
+
+        (0..self.r.len())
+            .map(|i| self.r[i] * t[i] + (E::ONE - self.r[i]) * (E::ONE - t[i]))
+            .fold(E::ONE, |acc, term| acc * term)
+    }
+
+    /// Computes $EQ((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1}))$ for all
+    /// $(y_0, ..., y_{ν - 1}) ∈ {0 , 1}^ν$ i.e., the Lagrange kernel at $r = (r_0, ..., r_{ν - 1})$.
+    pub fn evaluations(&self) -> Vec<E> {
+        compute_lagrange_basis_evals_at(&self.r)
+    }
+
+    /// Returns the evaluations of
+    /// $((y_0, ..., y_{ν - 1}) ↦ EQ^{~}((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1})))$
+    /// over ${0 , 1}^ν$.
+    pub fn ml_at(evaluation_point: Vec<E>) -> MultiLinearPoly<E> {
+        let eq_evals = EqFunction::new(evaluation_point.clone()).evaluations();
+        MultiLinearPoly::from_evaluations(eq_evals)
+    }
+}
+
+// HELPER
+// ================================================================================================
+
+/// Computes the evaluations of the Lagrange basis polynomials over the interpolating
+/// set ${0 , 1}^ν$ at $(r_0, ..., r_{ν - 1})$ i.e., the Lagrange kernel at $(r_0, ..., r_{ν - 1})$.
+fn compute_lagrange_basis_evals_at<E: FieldElement>(query: &[E]) -> Vec<E> {
```

Comment:
If there is no straightforward way to parallelize this right now, let's add a `TODO` comment to make sure we know that this would be an important function to parallelize.

Reply:
Added